iree/samples/custom_modules/dialect/CMakeLists.txt

```diff
@@ -71,7 +71,7 @@ iree_cc_binary(
   DEPS
     iree::samples::custom_modules::dialect
     iree::tools::iree_opt_library
-    MLIROptMain
+    MLIRMlirOptLib
```

Comment:
From what I can tell, the CMake configuration here matched Bazel, but this PR changes it. Do we want that?

In Bazel, `iree/tools:iree_opt_library` depends on `@llvm-project//mlir:MlirOptLib` [(here)](https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/iree/tools/BUILD#L48-L88) then `iree/tools:iree-opt` [(here)](https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/iree/tools/BUILD#L90-L96), `iree/samples/custom_modules/dialect:custom-opt` [(here)](https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/iree/samples/custom_modules/dialect/BUILD#L82-L89), and any other `iree_cc_binary` targets depend on both `iree/tools:iree_opt_library` and `@llvm-project//mlir:MlirOptMain`. This setup is deliberate: libraries can depend on any combination of other libraries, but binaries must only link in a single file containing a `main()` function. In this case, mlir-opt.cpp has that [main function](https://github.com/llvm/llvm-project/blob/df71000d7d5df2eb310ba25ddbfd6ab33bef6c56/mlir/tools/mlir-opt/mlir-opt.cpp#L50). If you want to use `iree_opt_library` with your own `main()` function, you can - just don't include the `MlirOptMain` target in your deps.

In CMake, `MLIRMlirOptLib` is the equivalent library and `MLIROptMain` includes the `main()` function.

References:
* [Bazel MlirOptLib](https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/build_tools/bazel/third_party_import/llvm-project/overlay/mlir/BUILD.bazel#L1719-L1748)
* [Bazel MlirOptMain](https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/build_tools/bazel/third_party_import/llvm-project/overlay/mlir/BUILD.bazel#L1796-L1807)
* [CMake MLIRMlirOptLib](https://github.com/llvm/llvm-project/blob/df71000d7d5df2eb310ba25ddbfd6ab33bef6c56/mlir/tools/mlir-opt/CMakeLists.txt#L5-L16)
* [CMake MlirOptMain](https://github.com/llvm/llvm-project/blob/430bba2a0f39b543e66119761d1526d037229936/mlir/lib/Support/CMakeLists.txt#L20-L30)

Reply:
Actually the targets in Bazel and CMake are currently swapped, please see #569. So if a Bazel target depends on `@llvm-project//mlir:MlirOptMain`, the equivalent CMake target is `MLIRMlirOptLib` which includes the `main()`. So I think this change is correct.

I currently have a review pending to revise this in MLIR's CMake configuration: https://reviews.llvm.org/D73778

Reply:
Argh, I see...

I'm going to try building more of iree/modules/strings (see #642) with these changes to see if they help some issues we were seeing there, then I think we could go ahead with this.

Maybe each time the deps are used they should include a comment that they should be renamed? This is a pretty confusing situation for developers adding these dependencies to Bazel and CMake without knowing the state of the upstream configuration.

```
MLIRMlirOptLib # main(); TODO: replace with MLIROptMain after swapping names
MLIROptMain # lib; TODO: replace with MLIRMlirOptLib after swapping names
```

Reply:
Ah, there's a separate issue where ALWAYSLINK evaluation returns early for targets that are not yet defined: https://github.com/google/iree/blob/bc7b80018ddc76d05f3e0e9a9e13d988740039ab/build_tools/cmake/iree_cc_binary.cmake#L186-L189

iree/modules/strings is configured before iree/tools/, so `iree_opt_library` does not yet exist (so `MLIROptMain` and the other deps are not pulled in).

Reply:
Okay, I think we can make this change, but I'd like the libraries to be labeled to help explain the swap.

Reply:
Sure, added comments to reduce confusion.